bundlegen/core/bundle_processor.py

```diff
@@ -536,29 +537,174 @@ def _process_logging(self):
+        libc['sublibs'] = []
+        for sublib in sublibs:
+            libc['sublibs'].append(sublib['name'])
+            sublib['parentlib'] = libc['name']
+            sublib['apiversions'] = []
+
+    def _take_rootfs_lib(self, srclib, dstlib, api_info):
+        """ The lib version from image rootfs was choosen. Basically log here that it was choosen.
+            Also for any sublibs if present.
+        """
+        logger.debug(f"IMAGE version choosen: {dstlib}")
+        self.handled_libs.add(dstlib)
+        if api_info and api_info.get('sublibs'):
+            for sublib in api_info['sublibs']:
+                logger.debug(f"IMAGE version choosen: {sublib}")
+                self.handled_libs.add(sublib)
+
+    # ==========================================================================
+    def _mount_or_use_rootfs(self, srclib, dstlib):
+        """Determine to mount lib from host OR use the one inside bundle rootfs.
+           If lib exists in rootfs and apiversions info exists for it inside *_libs.json config
+           then we try to use that version of the lib that has most API versions defined inside.
+           If lib does not exist inside image rootfs then create mount bind.
+           If no API versions info exists then create mount bind.
+           If decision cannot be made because API versions disjunct, choose image rootfs version.
+
+        Args:
+            lib (string): lib name like /lib/libc.so.6
+            mount_if_not_on_rootfs (bool): if lib doesn't exist inside rootfs
+                                           add a mount
+        """
+        if dstlib in self.handled_libs:
+            #logger.debug(f"Already handled: {dstlib}")
+            return
+
+        api_info = None
+        if self.platform_cfg.get('libs'):
+            api_info = [x for x in self.platform_cfg['libs'] if x['name'] == dstlib]
+        if not api_info:
+            logger.debug(f"No api info found for {dstlib}")
+            self._take_host_lib(srclib, dstlib, None)
+            return
+
+        api_info = api_info[0]
+
+        # if this is a sublib then switch to logic of parentlib instead
+        if api_info.get('parentlib'):
+            self._mount_or_use_rootfs(api_info['parentlib'], api_info['parentlib'])
+            return
+
+        rootfs_filepath = os.path.join(self.rootfs_path, dstlib.lstrip('/'))
+        if not os.path.exists(rootfs_filepath):
+            logger.debug(f"Lib not inside image rootfs {dstlib}")
+            self._take_host_lib(srclib, dstlib, api_info)
+            return
+
+        if len(api_info['apiversions']) > 0:
+            version_defs_by_host_lib = set(api_info['apiversions'])
+            version_defs_by_rootfs_lib = set(ReadElf.retrieve_apiversions(rootfs_filepath))
+
+            # remove from rootfs and mount lib from host if
+            # host lib contains more or same version definitions
+            if (version_defs_by_host_lib >= version_defs_by_rootfs_lib):
```

Comment:
We should make this library dependency adding optional. So we need to put parameter stating that it would be done or not. If so then we can also use it to make one more option to have it here to have exact same version (if this is needed in future). So there should be 3 states:
1. No matching
2. Matching higher and same (default)
3. Same version only matching

Reply:
We could introduce a parameter to disable dependency chain walking. But when that option is used, you must ensure that the platform json contains all dependencies for the gfxLibs and plugins as well. Because in effect it will only add bind mounts for what is listed in there.

Then, about a parameter to influence the library matching logic.  From the 3 states you mention above it is not always clear what should happen. Perhaps it is best to indicate with the parameter what gets priority. So something like this:

1. Check api version definitions and try to choose good library from host or rootfs = current implementation

2. Give priority to host. In effect this is your option 1 which means: ignore api version info and do "no matching" attempt. Keep in mind it still walks dependencies. This will add mount binds and remove any corresponding libs from rootfs

3. Give priority to dac image rootfs: when the library is found inside rootfs then use it and do not add a mount bind.

Reply:
Added two params, also to docs as such:

`-n, --nodepwalking              Dependency walking and library matching is
                                  active by default. Use this flag to disable
                                  it. When enabled, the dependencies of all
                                  libs indicated in gfxLibs and
                                  pluginDependencies config, will
                                  automatically also be added to the bundle.
                                  Host or image version of library is decided
                                  by libmatchingmode parameter below. This
                                  logic can only work if a _libs.json file is
                                  present with libs and apiversions info.

  -m, --libmatchingmode [normal|image|host]
                                  normal: take most recent library i.e. with
                                  most api tags like 'GLIBC_2.4'.

                                  image: always take lib from image rootfs, if
                                  available in there.

                                  host: always take host lib and create mount
                                  bind. Skips the library from image rootfs if
                                  it was there.

                                  Default mode is 'normal'. When apiversion
                                  info not available the effect is the same as
                                  mode 'host'`